### PR TITLE
Refactor AuthorList

### PR DESCRIPTION
**Pull request by tobiasdiez**
- Refactor AuthorList by extracting Author class and move parsing logic to new AuthorListParser
- Rename a few methods in AuthorList
- Add a few tests taken from biber https://github.com/plk/biber/blob/dev/t/names.t (two of them actually fail, so maybe somebody with better understanding of the inner workings of AuthorList may fix them)

--

- [ ] Change in CHANGELOG.md described
- [x] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)

### CODE REVIEW

File: src/main/java/net/sf/jabref/logic/autocompleter/NameFieldAutoCompleter.java [hunk 1]
```diff
@@ -93,7 +94,7 @@ public void addBibtexEntry(BibEntry entry) {
             if (entry.hasField(fieldName)) {
                 String fieldValue = entry.getField(fieldName);
                 AuthorList authorList = AuthorList.getAuthorList(fieldValue);
-                for (AuthorList.Author author : authorList.getAuthorList()) {
+                for (Author author : authorList.getAuthorList()) {
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/logic/autocompleter/NameFieldAutoCompleter.java]: AuthorList.getAuthorList() sounds really redundant from naming.
Together with the next line it is confusing on the first read.
What about a simple  `getAuthors()`

File: src/main/java/net/sf/jabref/logic/xmp/XMPUtil.java [hunk 2]
```diff
@@ -626,11 +626,10 @@ private static void writeToDCSchema(XMPSchemaDublinCore dcSchema,
              */
             if ("author".equals(field)) {
                 String authors = resolvedEntry.getField(field);
-                AuthorList list = AuthorList.getAuthorList(authors);
+                AuthorList list = AuthorList.getAuthors(authors);
 
-                int n = list.size();
-                for (int i = 0; i < n; i++) {
-                    dcSchema.addCreator(list.getAuthor(i).getFirstLast(false));
```
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/xmp/XMPUtil.java]: So the one-liner to get a List <Auhtor> is

AuthorList.getAuthors(String).getAuthors()?

Maybe not a huge difference but I sort of liked the previous naming better...

File: src/main/java/net/sf/jabref/logic/xmp/XMPUtil.java [hunk 3]
```diff
@@ -626,11 +626,10 @@ private static void writeToDCSchema(XMPSchemaDublinCore dcSchema,
...
+                for (Author author : list.getAuthors()) {
```
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/xmp/XMPUtil.java]: At least the first call could still well be getAuthorList() as that is more accurate.

File: src/main/java/net/sf/jabref/model/entry/AuthorList.java [hunk 4]
```diff
@@ -768,18 +392,18 @@ public String getAuthorsLastFirst(boolean abbreviate, boolean oxfordComma) {
         }
 
         StringBuilder result = new StringBuilder();
-        if (size() > 0) {
+        if (getNumberOfAuthors() > 0) {
```
Comment (oscargus) [inline src/main/java/net/sf/jabref/model/entry/AuthorList.java]: Is there an `isEmpty()` (or corresponding) method? Maybe there should be.

**Comment by oscargus** [comment]
Looks good! Have you tried the formatter in bst and seen if it formats incorrectly as well? I think that we should try to use that as much as possible instead of trying to reinvent it.

The only major comment is that getAuthors return different types. I think it makes sense to keep getAuthorList (or use a different name).

File: src/main/java/net/sf/jabref/logic/autocompleter/NameFieldAutoCompleter.java (hunk 1 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/autocompleter/NameFieldAutoCompleter.java]: This is actually done in a latter commit.

File: src/main/java/net/sf/jabref/logic/xmp/XMPUtil.java (hunk 2 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/xmp/XMPUtil.java]: previously it was AuthorList.getAuthorList.getAuthorList...has also a duplication. Mhh...what do you think about renaming the first getAuthor to parse? I.e `AuthorList.parse(String).getAuthors`. That the parsing is actually cached shouldn't be a worry for the name.

File: src/main/java/net/sf/jabref/logic/autocompleter/NameFieldAutoCompleter.java (hunk 1 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/autocompleter/NameFieldAutoCompleter.java]: OK! Sounds good! I use some tool on the phone to check it so maybe that has
problems. I can for example not see your comment in the tool...

File: src/main/java/net/sf/jabref/model/entry/AuthorList.java (hunk 4 above)
Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/model/entry/AuthorList.java]: Yes there is an isEmpty...should be used here you are right.

File: src/main/java/net/sf/jabref/logic/xmp/XMPUtil.java (hunk 2 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/xmp/XMPUtil.java]: I think I introduced the second getAuthorList... Didn't think of it then...

Or getAuthorList().getListOfAuhtors() which in some sense is more
self-explanatory but maybe also open for more confusion. Your suggestion
works for me, so as long as they are different I'm OK with it.

Comment (tobiasdiez) [inline src/main/java/net/sf/jabref/logic/xmp/XMPUtil.java]: I now renamed it to `parse` which reads quite well in my opinion.

Comment (oscargus) [inline src/main/java/net/sf/jabref/logic/xmp/XMPUtil.java]: :+1:

**Pull request merged**